forte/data/extractor/base_extractor.py

```diff
@@ -0,0 +1,205 @@
+            "vocab": self.vocab.state if self.vocab else None,
+        }
+
+    @classmethod
+    def from_state(cls, state: Dict):
+        config = {
+            "vocab_method": state["vocab_method"],
+            "need_pad": state["need_pad"],
+            "vocab_use_unk": state["vocab_use_unk"],
+            "entry_type": state["entry_type"]
+        }
+        obj = cls(config)
+        if "vocab" in state and state["vocab"] is not None and \
+            state["vocab"] != "None":
+            obj.vocab = Vocabulary.from_state(state["vocab"])
+        else:
+            obj.vocab = None
+        return obj
+
+    @property
+    def entry_type(self) -> Type[Annotation]:
+        return self.config.entry_type
+
+    @property
+    def vocab_method(self) -> str:
+        return self.config.vocab_method
+
+    def get_pad_value(self) -> Union[None, int, List[int]]:
+        if self.vocab:
+            return self.vocab.get_pad_value()
+        else:
+            return None
+
+    def items(self) -> Iterable[Tuple[Hashable, int]]:
+        assert self.vocab
+        return self.vocab.items()
+
+    def size(self) -> int:
+        assert self.vocab
+        return len(self.vocab)
+
+    def add(self, element: Hashable):
+        assert self.vocab
+        return self.vocab.add_element(element)
+
+    def has_element(self, element: Hashable) -> bool:
+        assert self.vocab
+        return self.vocab.has_element(element)
+
+    def element2repr(self, element: Hashable) -> Union[int, List[int]]:
+        assert self.vocab
+        return self.vocab.element2repr(element)
+
+    def id2element(self, idx: int) -> Any:
+        assert self.vocab
+        return self.vocab.id2element(idx)
+
+    def get_dict(self) -> Dict[Hashable, int]:
+        assert self.vocab
+        return self.vocab.get_dict()
+
+    def predefined_vocab(self, predefined: Union[Set, List]):
+        """This function will add elements from the
+        passed-in predefined set to the vocab. Different
+        extractors might have different strategies to add
+        these elements. Override this function if necessary.
+
+        Args:
+            predefined (Union[Set, List]): A set or list contain
+                elements to be added into the vocabulary.
+        """
+        for element in predefined:
+            self.add(element)
+
+    def update_vocab(self, pack: DataPack,
+                    instance: Annotation):
+        """This function will extract the feature from
+        instance and add element in the feature to vocabulary.
+
+        Args:
+            pack (Datapack): The datapack that contains the current
+                instance.
+            instance (Annotation): The instance from which the
+                extractor will extractor feature.
+        """
+        raise NotImplementedError()
+
+    def extract(self, pack: DataPack,
+                instance: Annotation) -> Feature:
+        """This function will extract feature from
```

Comment:
"Implement this function to perform feature extraction. This functions should yield one instance at a time from the pack."

Should this be implemented as a Iterator?

Reply:
1. The above docstring will be added.
2. Yeah, I think this is a design choice. Currently, the responsibility of looping over all the instance in a pack is handled by the `train_preprocessor`. The advantage of our implementation is that we can get the feature for a specific instance in a pack, while if we implement it as an iterator, we do not have the flexibility to access the feature at arbitrary position.